File: Study Plan/Algorithm I/Population_next_Right_Pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None:
            return root
        
        self.traverse(root, None, None)
        return root
    
    def traverse(self, node, parent, goLeft):
        if node is None:
            return 
        
        left, right = node.left, node.right
        self.traverse(left, node, True)
        
        if parent is None:
            # for root
            node.next = None
        elif goLeft:
            node.next = parent.right
        else:
            if parent.next is None:  
                node.next = None
            else:
                node.next = parent.next.left
        
        self.traverse(right, node, False)
        

File: Study Plan/Algorithm I/Population_next_Right_Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
	
        if(!root)
            return NULL;
			
        queue<Node*> q;
        q.push(root);
		
        while(!q.empty()){
            int sz=q.size();
			
            while(sz--){
                Node* p=q.front();
                q.pop();
				
                if(sz!=0)
                    p->next=q.front();

                if(p->left)
                    q.push(p->left);
                if(p->right)
                    q.push(p->right);
            }
        }
        return root;
    }
};